healthchain/pipeline/basepipeline.py

```diff
@@ -0,0 +1,463 @@
+            position (str): The position where the component should be added.
+                            Valid values are 'first', 'last', 'after', 'before', or 'default'.
+            reference (str, optional): The name of the reference component when using 'after' or 'before' positions.
+
+        Raises:
+            ValueError: If an invalid position is provided or if a reference is required but not provided.
+
+        This method handles the insertion of a new component into the pipeline based on the specified position:
+        - 'first': Inserts the component at the beginning of the pipeline.
+        - 'last' or 'default': Appends the component to the end of the pipeline.
+        - 'after' or 'before': Inserts the component relative to a reference component.
+
+        For 'after' and 'before' positions, a reference component name must be provided.
+        """
+        if position == "first":
+            self._components.insert(0, new_component)
+        elif position in ["last", "default"]:
+            self._components.append(new_component)
+        elif position in ["after", "before"]:
+            if not reference:
+                raise ValueError(
+                    f"Reference must be provided for position '{position}'."
+                )
+            offset = 1 if position == "after" else 0
+            self._insert_relative_position(new_component, reference, offset)
+        else:
+            raise ValueError(
+                f"Invalid position '{position}'. Must be 'first', 'last', 'after', 'before', or 'default'."
+            )
+
+    def _insert_relative_position(self, component, reference, offset):
+        """
+        Insert a component relative to a reference component in the pipeline.
+
+        Args:
+            component (PipelineNode): The component to be inserted.
+            reference (str): The name of the reference component.
+            offset (int): The offset from the reference component (0 for before, 1 for after).
+
+        Raises:
+            ValueError: If the reference component is not found in the pipeline.
+        """
+        ref_index = next(
+            (i for i, c in enumerate(self._components) if c.name == reference), None
+        )
+        if ref_index is None:
+            raise ValueError(f"Reference component '{reference}' not found.")
+
+        self._components.insert(ref_index + offset, component)
+
+    def remove(self, component_name: str) -> None:
+        """
+        Removes a component from the pipeline.
+
+        Parameters:
+        - component_name (str): The name of the component to be removed.
+
+        Returns:
+        None
+
+        Logs:
+        - WARNING: If the component is not found in the pipeline or fails to be removed.
+        """
+        # Check if the component exists in the pipeline
+        if not any(c.name == component_name for c in self._components):
+            logger.warning(f"Component '{component_name}' not found in the pipeline.")
```

Comment:
No real issue here but I'm just wondering whether this edge case should fail louder? 

Specifically I see this warning being triggered by typos on r some underlying mistake. I think the user in that case would want to be explicitly thrown an error to help them catch the mistake.

Reply:
good point